File: combinations.py
class Solution:
    # @return a list of lists of integers
    def combine(self, n, k):
    	def helper(choices,visited,start):
    		#if choices == 1:
    		#	return [[i+1] for i in range(n)]
    		if choices == 0:
    			return [ [] ]
    		result = []
    		for i in range(start,n):
    			if not visited[i]:
    				visited[i] = True
    				temp = helper(choices-1, visited, i+1)
    				#print temp, start, n
    				
    				for j in temp:
    					t = [i+1] + j
    					result.append(t)
    				visited[i] = False
    		return result

    	visited = [False for i in range(n)]
    	return helper(k, visited, 0)

s = Solution()
print s.combine(4,2)

File: JAVA/combinations.java
public class Solution {
    public List<List<Integer>> combine(int n, int k) {
        Boolean[] visited = new Boolean[n];
        for (int i = 0; i < visited.length; i++)
            visited[i] = false;
        return dfs(0, visited, k, n);
    }
    
    public List<List<Integer>> dfs(int start, Boolean[] visited, int remainK, int n){
        ArrayList<List<Integer>> result = new ArrayList<List<Integer>>();
        List<List<Integer>> rest;
        if (remainK == 0){
            result.add(new ArrayList<Integer>());
            return result; 
        }
        for (int j=start; j<n; j++){
            if (visited[j]) 
                continue;
            visited[j] = true;
            rest = dfs(j+1, visited, remainK-1, n);
            for(List<Integer> lst: rest){
                lst.add(0, j+1);
                result.add(lst);
            }
            visited[j] = false;
        }
        return result;
    }
}